contracts/cw-names-registry/src/contract.rs

```diff
@@ -0,0 +1,256 @@
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::Receive(wrapped) => execute_receive(deps, env, info, wrapped),
+        ExecuteMsg::UpdateConfig {
+            new_payment_token_address,
+            new_admin,
+            new_payment_amount,
+        } => execute_update_config(
+            deps,
+            env,
+            info,
+            new_payment_token_address,
+            new_admin,
+            new_payment_amount,
+        ),
+        ExecuteMsg::Revoke { name } => execute_revoke(deps, env, info, name),
+        ExecuteMsg::Reserve { name } => execute_reserve(deps, env, info, name),
+        ExecuteMsg::TransferReservation { name, dao } => {
+            execute_transfer_reservation(deps, env, info, name, dao)
+        }
+    }
+}
+
+pub fn execute_receive(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    wrapped: Cw20ReceiveMsg,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+
+    // We only take payments from our specified token
+    if info.sender != config.payment_token_address {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    // TODO: Is there a way we can verify this is a DAO
+    let sender = wrapped.sender;
+    let amount = wrapped.amount;
+    let msg: ReceiveMsg = from_binary(&wrapped.msg)?;
+
+    match msg {
+        ReceiveMsg::Register { name } => register_name(deps, env, sender, amount, name),
+    }
+}
+
+pub fn register_name(
+    deps: DepsMut,
+    _env: Env,
+    sender: String,
+    amount: Uint128,
+    name: String,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    if amount < config.payment_amount {
+        return Err(ContractError::InsufficientFunds {});
+    }
+
+    // this is the DAO
+    let sender = deps.api.addr_validate(&sender)?;
+
+    if RESERVED_NAMES.has(deps.storage, name.clone()) {
+        return Err(ContractError::NameReserved {});
+    }
+
+    if NAME_TO_DAO.has(deps.storage, name.clone()) {
+        return Err(ContractError::NameAlreadyTaken {});
+    }
+
+    if DAO_TO_NAME.has(deps.storage, sender.clone()) {
+        return Err(ContractError::NameAlreadyRegistered {});
+    }
+
+    NAME_TO_DAO.save(deps.storage, name.clone(), &sender)?;
+    DAO_TO_NAME.save(deps.storage, sender, &name)?;
+
+    Ok(Response::new())
+}
+
+pub fn execute_update_config(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    new_payment_token_address: Option<String>,
+    new_admin: Option<String>,
+    new_payment_amount: Option<Uint128>,
+) -> Result<Response, ContractError> {
+    let mut config = CONFIG.load(deps.storage)?;
+
+    if info.sender != config.admin {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let new_payment_token_address =
+        new_payment_token_address.unwrap_or_else(|| config.payment_token_address.to_string());
+    let payment_amount = new_payment_amount.unwrap_or(config.payment_amount);
```

Comment:
Looks like we're not checking whether payment is 0 here. That's fine - as long as it's consistent with `instantiate`.